[Rex Yong Jin Xiang] iP

rexyyong opened the pull request.

rexyyong committed:
Level 0: Rename, Greet, Exit

rexyyong committed:
Add Level-1, Echo

Chatbot only displays welcome message.

Echo method is required within chatbot.

Echo method created which echoes text from users. It exits when string bye, case insensitive is detected.

rexyyong committed:
Add Level-2, (Add, List)

Omoh class can only do echo

Functionality of adding tasks and displauing tasks to users is required

Create Class List that stores array of Strings of size 100. Create methods that allow tasks to be stored and retrieved from List

Create Class list so that code becomes neater

rexyyong committed:
Add new Class Task

Edit Class List such that it now has Class Task added to it.

Functionality of Class Task required such that class variables "String description" and "Boolean isDone" can be stored within the object, telling user what task it is and whether it is completed or not.

rexyyong committed:
Add mark task done feature

mark i, where i is an integer will cause the task with index i to be marked as done

Add extractTaskNumber method in class Omoh to determine the task number user wants to mark
Add markAsDone, markAsNotDone, printMarkAsDone methods in Class List to change done state of task and also to print out the state of tasks

rexyyong committed:
Add unmark task feature

Modify method readUserInput in Class Omoh such that it can detect whether user typed in mark or unmark.

Add extractTaskNumber method in Class Omoh such that according to task numer indicated by user, program can decide which task to mark or unmark

Add method modifyDoneState in Class Omoh to mark task in list as done or undone

Add method printMarkTask in Class List to print out which task was marked or unmarked by user

rexyyong committed:
Do codingstandards and format code

rexyyong committed:
Add store and print task and deadline feature

Use inheritance to create class Deadline
Create condition to look for "deadline" keyword in input and extract the task and deadline
print out task and deadline after it is added

rexyyong committed:
Remove echo method from Class omoh

rexyyong committed:
Add todo type of task

Readuserinput in Omoh class detects if todo keyword is typed
Add in addTodo function that extracts the description after todo keyword
Add in printTodoMessage function that prints out what was added

rexyyong committed:
Add function to add in events

Add Class Event that stores description, to and from
Add condition in Class Omoh to detect event keyword in input
Add function in Class List that helps to add event, and print out add event message to CLI

rexyyong committed:
Modify printing of all tasks when list is typed

getAllTasks function in Class List shows type of Task, as well as doneness of task when "list" keyword is typed by user
Add member String Type in Class Task, Class Event, Class Deadline to store the type of task

rexyyong committed:
Fix listing of tasks

Modify function getAllTasks() in Class List so that based on type of task, the memebrs will be printed
Fix formating for messages printed when tasks added

rexyyong committed:
Put function extractTaskNumber and function modifyDoneState under Class Task

Make code more readable and neater

rexyyong committed:
Create printWelcomeMessage function in Class Omoh

rexyyong committed:
Shift function addEventMessage from Class List to Class Event

rexyyong committed:
Shift printDeadline function from Class List to Class Deadline

rexyyong committed:
Shift function printMarkTask from Class List to Class Task

rexyyong committed:
Remove redundant function markAsDone and markAsNotDone in Class List

rexyyong committed:
Shift function addEvent from Class List to Class Event

rexyyong committed:
Shift function addDeadline from Class LIst to Class Deadline

rexyyong committed:
Shift function getAllTasks from Class List to CLass Task

rexyyong committed:
Shift function addTask and function printAddedTask from Class List to Class Task

rexyyong committed:
Remove redundant line in Class List

rexyyong committed:
Add Class Todo

Shift function addTodo and function printTodoMessage from CLass List to CLass Todo

rexyyong committed:
Delete Class List, shift everything to Class Task

Organise everything to Class Task so that code is neater. Class List is redundant because Class Task can manage the functionalities

rexyyong committed:
Remove indentation printing of messages on CLI

CLI now displays all texts without indentation. Remove indentation such that code is a bit neater

rexyyong committed:
Remove redundant code in Class Task

rexyyong committed:
Add toString method in Class Task

Method allows printing to be neater

rexyyong committed:
Add toString() method

Add method to Class Task, Class Todo, Class Event, Class Deadline
Include override for toString method

rexyyong committed:
Edit some functions in Class Task to be neater

rexyyong committed:
Add Automated testing

check Todo, Event, Deadline
check mark
check bye

ChongXern commented on src/main/java/List.java:
```diff
@@ -0,0 +1,31 @@
+        int serialNumber = 1;
+        for (int i = 0; i < totalTasks; i++) {
+            System.out.print("     ");
+            System.out.println(serialNumber + ". " + tasks[i]);
```
Can consider adding the spaces together with the println for better readability as you have earlier

ChongXern commented on src/main/java/Omoh.java:
```diff
@@ -16,10 +16,8 @@ public static void main(String[] args) {
-            System.out.print(" ");
-        }
+    public static void printHorizontalLine() {
+        System.out.print("    ");
```
Great job in including spaces to make code more user friendly. Could avoid DRY rule in by possibly including it in your class as a part of a method

ChongXern commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+public class Task {
```
Good coding standards applied overall.

ChongXern commented on src/main/java/List.java:
```diff
@@ -31,6 +31,24 @@ public void printAddedTask() {
+        System.out.print("     ");
+        System.out.println("Got it. I've added this task:");
+        System.out.print("       ");
+        System.out.println("[D][ ]" + description.description + " (by: " + description.by + ")" );
```
The [D] part could be added as a boolean in the method so it could be easier to adjust in the future

ChongXern commented on src/main/java/Event.java:
```diff
@@ -37,4 +37,9 @@ public static void addEvent (String input) {
         Event.addEventMessage(extractedInfo);
     }
+
+    @Override
```
All in all, the Java coding standards used throughout have been exemplary

ChongXern reviewed (commented):
The code here is clearly well thought out in terms of Java coding standards and highly commendable to learn from. I've gained valuable insights from this code, which I highly appreciate and will learn from.

chenhowy commented on src/main/java/Omoh.java:
```diff
@@ -0,0 +1,79 @@
+        String line;
+        Scanner in = new Scanner(System.in);
+        line = in.nextLine();
+        while (!line.equalsIgnoreCase("bye")) {
```
Consider adding the `line = in.nextLine();` after the if/else instead of writing it in every case.

rexyyong replied:
thanks for the input, have implemented this!

chenhowy commented on src/main/java/Omoh.java:
```diff
@@ -0,0 +1,79 @@
+    //if list is typed, list out all tasks
+    //if mark or unmark is typed, mark or unmark the task in tasklist
+    //if none of the above, add the typed input to task list
+    //If bye is typed, function exits
```
Clear description of how input is handled.

chenhowy commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,101 @@
+    protected String description;
+    protected boolean isDone;
+
+    protected String type;
```
What is this 'type' used for? I may be wrong, but I don't see where this variable is being used.

rexyyong replied:
Oh yes you're not wrong, it's not used for anything anymore. I forgot to remove it but basically previously I used type to store whether the object is Task, Deadline, or Event

chenhowy commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+    public static void printDeadline (Deadline description) {
+        Omoh.printHorizontalLine();
+        System.out.println("Got it. I've added this task:");
+        System.out.println("[D][ ] " + description.description + " (by: " + description.by + ")" );
```
Perhaps the .type variable can be used here to replace "[D]", allowing for easier change of type letters in the future.

rexyyong replied:
Oh yes thats a good idea. Thank you for the insight!

rexyyong committed:
Add in if else case to handle empty input case

rexyyong committed:
Add EmptyTodoException

rexyyong committed:
Do exception from IllegalEventInput, ensure that Event keyed in properly

rexyyong committed:
Make IllegalEventInput exception more robust

rexyyong committed:
Create exception for deadline

rexyyong committed:
Add exception handling for mark and unmark

rexyyong committed:
Make exceptions for mark and unmark task more robust

rexyyong committed:
Update test case to test mark exceptions

rexyyong committed:
Merge branch 'branch-Level-5'

rexyyong committed:
Create packages, fix automated test

rexyyong committed:
Improve automated testing

rexyyong committed:
Merge branch 'branch-A-Packages'

rexyyong committed:
Create Jar for tutorial, contains up to LEVEL 5

nichyjt commented on src/main/java/omoh/Omoh.java:
```diff
@@ -0,0 +1,103 @@
+
+        while (!line.equalsIgnoreCase("bye")) {
+            if (line.trim().isEmpty()) {
+                System.out.println("Please enter a non empty Input!");
+            } else if (line.equalsIgnoreCase("list")) {
+                printAllTasks();
+            } else if (line.startsWith("deadline")) {
+                Deadline.addDeadline(line);
+            }
+            else if (line.startsWith("todo")) {
+                try {
+                    Todo.addTodo(line);
+                } catch (EmptyTodoException e) {
+                    System.out.println("Todo cannot be empty! Please key in input in this format [Todo] [task] " +
+                            "example: todo hang clothes");
+                }
+            }
+            else if (line.startsWith("event")) {
+                Event.addEvent(line);
+            }
+            else if (line.startsWith("mark") || line.startsWith("unmark")) {
+                try {
+                    int taskNumber = Task.extractTaskNumber(line);
+                    Task.modifyDoneState(taskNumber, line);
+                    Task.printMarkTask(taskNumber, line);
+                } catch (NumberFormatException e) {
+                    System.out.println("Please enter valid number without alphabets after mark or unmark " +
+                            "Example: mark 1");
+                } catch (NullPointerException e) {
+                    System.out.println("Please enter valid number within boundaries of list. " +
+                            "Example: if list only has 2 items, dont enter beyond 2");
+                } catch (EmptyTaskNumberException e) {
+                    System.out.println("Please enter a number after mark or unmark " +
+                            "Example: mark 1");
+                }
+            } else {
+                Task.addTask(line);
+                Task.printAddedTask();
+            }
+            line = in.nextLine();
```
I think this is a good attempt at making the code SLAP. I think there may be room for improvement to make this section SLAP more.

Here are some guiding questions...
1. What parts of the code may be repetitive? i.e. virtually the same for each if/else block you have
2. What part of the code does a particular function that can be isolated into a function? (like printing tasks?)

nichyjt commented on src/main/java/omoh/customexceptions/EmptyTodoException.java:
```diff
@@ -0,0 +1,4 @@
+package omoh.customexceptions;
+
+public class EmptyTodoException extends Exception{
```
Formatting error at the curly brace!

nichyjt commented on src/main/java/omoh/customexceptions/EmptyTaskNumberException.java:
```diff
@@ -0,0 +1,4 @@
+package omoh.customexceptions;
+
+public class EmptyTaskNumberException extends Exception {
```
I am not sure why the exception class here is completely empty? Shouldn't there be some kind of functionality?

nichyjt commented on src/main/java/omoh/tasktypes/Task.java:
```diff
@@ -0,0 +1,97 @@
+    //method that modifies whether task is done or not done, depending on keyword mark or unmark detected
+    public static void modifyDoneState(int taskNumber, String input) {
+        //only executes if taskNumber is valid
+        if (taskNumber != -1) {
```
You could possibly refactor this into a guard clause to avoid deep nesting! Think about how you can apply the idea of guard clauses thoughout your codebase to make the [happy path more prominent](https://nus-cs2113-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent).

Also, there are many magic literals throughout the codebase. Do refactor them! For example, it is not entirely clear what `-1` means here.

nichyjt reviewed (commented):
Hi Rex,

Your code is generally well structured. Good job!

There are some other specific comments below. Try and fix them before the next tutorial!

For your consideration: There is a way to autoformat your code automatically every time you Ctrl-S on intellij which will save you much pain. :)

rexyyong committed:
Implement Collections with ArrayList

Modify code to shift from conventional arrays to using collection ArrayList
Edit Parent Class Task to include public static ArrayList<Task> tasks
Edit functions in Parent Class Task to ArrayList<Task> tasks
Edit functions in subclass Deadline, Event, Todo to use ArrayList<Task> tasks as well

Edit exception in Omoh.java to catch IndexOutOfBoundsException instead of NullPointerException

rexyyong committed:
Add Delete functionality

Edit Omoh Class to include case where user types in delete
Edit Task.java to handle delete case
Modify function modifyDoneState to modifyDoneStateOrDelete in Class Task
Add printDeleteTask function in Class Task
Edit EXPECTED.txt and inputs.txt to test for delete

rexyyong committed:
Add printed output that says how many tasks are left in list after deletion

Edit EXPECTED.txt to match the changes

rexyyong committed:
Add function to create file under ip/data/output

rexyyong committed:
Add function to write to output.txt

writeToFile function called when bye is called

Under Class Task, create writeToFile function that transfers list of tasks to output.txt

Create getter in Class Event to allow event details to be accessed in writeToFile function

rexyyong committed:
Add read output.txt and initialise list function

Function readFile in Class Task reads from output.txt at start of programme.

rexyyong committed:
Add function to mark task based on output.txt

rexyyong committed:
Add read and execute unmark function

rexyyong committed:
Handle folder not found case

rexyyong committed:
Add exception to handle corrupted file

rexyyong committed:
Merge branch 'branch-Level-7'

rexyyong committed:
Fix test case with level 6 merged in

Need to delete output.txt at test folder everytime test is run

rexyyong committed:
Merge branch 'branch-Level-6'

rexyyong committed:
Fix writeToFile method in Class Task to use ArrayList instead

rexyyong committed:
Create method createFileDirectory in Class Task.

Extract method from main to Class Task to make code neater

rexyyong committed:
Add comment for createFileDirectory method in Task class

rexyyong committed:
Extract createOutputFile from class Omoh to class Task

rexyyong committed:
Add comment in class Omoh

rexyyong committed:
Shift getStatusIcon to another line

rexyyong committed:
Extract processFileText method from readfile method

rexyyong committed:
Add exception for readFile method. Add comment for readFile method

rexyyong committed:
Add space to fix format at curly brace

rexyyong committed:
Remove redundant if statement

rexyyong committed:
Add message to print out in EmptyTodoException. Print StackTrace in catch block

rexyyong committed:
Add message to print in EmptyTaskNumberException. Call stack trace in catch block

rexyyong committed:
Merge branch 'refactorWeek6'

rexyyong committed:
Create new Class Ui. Shift method readUserInput to Class Ui

rexyyong committed:
Remove unnecessary import statements in Class Omoh

rexyyong committed:
Shift method printAllTasks from Class Omoh to Class Task

rexyyong committed:
Create Class Storage. Shift method createFileDirectory into Class Storage

rexyyong committed:
Rename createOutputFile to createOrReadOutputFile

rexyyong committed:
Shift CreateOrReadOutputFile method from Class Task to Class Storage

rexyyong committed:
Shift method readFile from Class Task to Class Storage

rexyyong committed:
Shift method processFileText from class Task to Class Storage

rexyyong committed:
Remove redundant import statements from Class Task

rexyyong committed:
Update tasks array from public to protected

rexyyong committed:
Create Class Parser

rexyyong committed:
Shift extractTaskNumber method from Class Task to Class Parser

rexyyong committed:
Extract method exyracyTaskAndDueDate from Class Deadline to Class Parser

rexyyong committed:
Extract method extractEvent from Class Event to Class Parser

rexyyong committed:
Update output.txt file

rexyyong committed:
Add findMatchingTasks method in Class Task. Call method in Class Ui

rexyyong committed:
Add extractKeyword method in Class Parser to use in findMatchingTasks method in Class Task

rexyyong committed:
Add EmptyFindException exception to handle empty find input

rexyyong committed:
Add searchTasks method in Class Task to use in findMatchingTasks method

rexyyong committed:
Add printFindResults method to use in findMatchingTasks method in Class Task

rexyyong committed:
Update input.txt to test find method

Test with tasks marked
Test with tasks unmarked
Test empty find input

rexyyong committed:
Add javadoc comments in Class Omoh

rexyyong committed:
Add javadoc comments for Class Parser

rexyyong committed:
Add javadoc for Class Event

rexyyong committed:
Add javadoc for Class Ui

rexyyong committed:
Add javadoc for Class Storage

rexyyong committed:
Add Javdoc for Class Todo

rexyyong committed:
Remove unnecessary lines in Class Deadline

rexyyong committed:
Remove another uncessary line in Class Deadline

rexyyong committed:
Add javadoc for Class Deadline

rexyyong committed:
Add javadoc to Class Task

rexyyong committed:
Merge pull request #1 from rexyyong/branch-Level-9

Branch level 9: Find

rexyyong committed:
Change output.txt file

rexyyong committed:
Merge branch 'master' of https://github.com/rexyyong/ip

rexyyong committed:
Merge branch 'master' into A-JavaDoc

# Conflicts:
#	src/main/java/omoh/Parser.java

rexyyong committed:
Merge pull request #2 from rexyyong/A-JavaDoc

A-JavaDoc, add in JavaDoc comments

rexyyong committed:
Update Readme with Introduction

rexyyong committed:
Add in features heading in ReadMe

rexyyong committed:
Add in feature descriptions in readme

rexyyong committed:
Update markdown in readme

rexyyong committed:
Update description of features in readme

rexyyong committed:
Add blank lines between features

rexyyong committed:
Add content Page

rexyyong committed:
Update content page in Readme to include features

rexyyong committed:
Add emojis to readme

rexyyong committed:
Add bold to 2 headings

rexyyong committed:
Add line breaks in features

rexyyong committed:
Merge pull request #3 from rexyyong/UserGuide

User Guide for Omoh

rexyyong committed:
Remove method addTask from Class Task

rexyyong committed:
Remove method printAddedTask from Class Task

rexyyong committed:
Add message to prompt users to type in valid command

rexyyong committed:
Add test to check if program detects illegal command

rexyyong committed:
Merge branch 'remove_general_task'

rexyyong committed:
Update README.md in docs folder

rexyyong committed:
Update README.md to remove emojis
